qiskit/transpiler/passes/synthesis/qubit_tracker.py

```diff
@@ -0,0 +1,123 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2024.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""A qubit state tracker for synthesizing operations with auxiliary qubits."""
+
+from __future__ import annotations
+from collections.abc import Iterable
+from dataclasses import dataclass
+
+
+@dataclass
+class QubitTracker:
```

Comment:
I really like this new class that very clearly exposes the functionality of how we track clean and dirty ancilla qubits. A couple of questions:
* Do we really need `self.qubits`, i.e. wouldn't it always be the same as the union of `self.clean` and `self.dirty`? Do we need this for the `borrow` method to be deterministic? Or for something else as well?
* We should probably specify the numbers of required clean and dirty qubits for the `borrow` method, i.e. have the method be something like `borrow(self, num_clean_qubits, num_dirty_qubits, active_qubits)`?
* When would a method `drop` be used? 
* It might be nice to add some example on using various methods in this class.

Reply:
* Yes, we need it to be deterministic, but this is a choice we have. We could also change `borrow` to sort the qubits to make the output deterministic.
* We could add that, but currently it would remain unused since HLS plugins have no way of telling how many ancillas of which _kind_ they are using -- only how many. I'm slightly leaning towards adding this feature once we can use it, but I'm open to changing my mind 🙂 
* `drop` is used for annotated controlled operations. There, we do not want to include the control qubits when synthesizing the base op, hence we do `tracker.drop(control_qubits)` 🙂 
* Sure, I can add some 👍🏻